### DISCUSSION TITLE
Migrate from v3 to v5

### DISCUSSION THREAD

**Discussion by qetza** [qetza/replacetokens-task#33] [Announcements]
The only difference between v3 and v5 is the new input `tokenPattern`.

When migrating from v3 to v5 use the following table to set this input based on your current values of `tokenPrefix` and `tokenSuffix`:

| tokenPrefix | tokenSuffix | tokenPattern | Comments |
|-|-|-|-|
|  |  |  |  if not set leave default values |
| `#{` | `}#` | `default` | you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `$(` | `)` | `azpipelines` | you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `{{` | `}}` | `doublebraces` | you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `__` | `__` | `rm` | **only for v5**, you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `__` | `__` | `doubleunderscores` | **only for v6**, you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `#{` | `}` | `octopus` | you can remove `tokenPrefix` and `tokenSuffix` inputs |
| `${{` | `}}` | `githubactions` | **only for v6**, you can remove `tokenPrefix` and `tokenSuffix` inputs |
| other | other | `custom` | keep `tokenPrefix` and `tokenSuffix` inputs |

**Comment by andriikut** [comment]
I want to add point: if you're migrating from 3 to 5, and using some custom pattern, it is critical to add 
  `        tokenPattern: custom `
  to inputs, otherwise v5 task will not replace tokens, but will be succeeded 
  version3 was worked without tokenPattern: custom

**Comment by parshOne** [comment]
Can u please let me know how long would v5 will be around and supported for?

**Reply by qetza** [reply]
Hi @parshOne,
I don't intend to remove any version so all versions will still be available in the extension.

As for support i don't think the v5 version will be deprecated soon, i'll still update it to support new node version but for new features it will depend it will depend on the time needed to implement them as for new features i will privilege version 6 as it's using a common library which allows the extension to be available as a CLI and a GitHub Action.

In any case as the code is open-source anyone can fork it and update it as needed :smile:

**Comment by jase52476** [comment]
Can someone please explain how token pattern azpipelines would work? If I am using an Azure CICD pipeline, I connect to a keyvault and the variables are accessible with the same format : $(variableKey) is this option for tokenPattern just for consistency or is it actually doing something different? Currently I'm using the doublebraces, and it's working fine, so I'm wondering if the azpipelines option has any special functionality. 

Thank you.

**Reply by qetza** [reply]
> Can someone please explain how token pattern azpipelines would work? If I am using an Azure CICD pipeline, I connect to a keyvault and the variables are accessible with the same format : $(variableKey) is this option for tokenPattern just for consistency or is it actually doing something different? Currently I'm using the doublebraces, and it's working fine, so I'm wondering if the azpipelines option has any special functionality.
> 
> Thank you.

Hi @jase52476,
In your azure pipelines YAML you can use the syntax `$(variableKey)` so that the pipeline engine replaces those by the variable value before executing the task so replacement is done on the YAML file defining your pipeline.

The replace tokens task will do variable replacement in the **content** of any text based files. Using the pattern `azpipelines` for the task is just to reuse the same format as the one use for variable replacement inside the YAML.

Hope this answers your question.

**Reply by jase52476** [reply]
@qetza  thank you, this is what I was guessing--it was for convenience so you can use the same format.  But I wasn't sure if there was some other integration with the pipeline that I was missing. I have a situation where I need some replacements in my Helm chart values files before running the HelmDeploy@1 task, so was using the double braces or double underscore, but maybe switching to the `azpipelines` option might be better for consistency.

Thanks for clearing it up for me.

**Comment by parshOne** [comment]
Hi Jason,
The primary purpose of this task in azure pipeline is to find specific
token patterns (e.g., #{MyVariable}#, __MyVariable__) within text-based
file.
For the case which you are using right now it doesn't matter much but let's
say you have a configuration file for your application that contains
placeholders for values that change per environment (e.g., database
connection strings, API endpoints, feature flags). You wouldn't want to
hardcode these values in your source code. If here you would have used
azure key vault then you have to change values again and again for
different env.

Thank you.
Parvesh
